File: picocli-examples/src/main/java/picocli/examples/interactive/PasswordDemo.java
package picocli.examples.interactive;

import picocli.CommandLine;
import picocli.CommandLine.Model.CommandSpec;
import picocli.CommandLine.Option;
import picocli.CommandLine.ParameterException;
import picocli.CommandLine.Spec;

import java.io.File;

public class PasswordDemo implements Runnable {
    @Option(names = "--password:file")
    File passwordFile;

    @Option(names = "--password:env")
    String passwordEnvironmentVariable;

    @Option(names = "--password", interactive = true)
    String password;

    @Spec
    CommandSpec spec;

    public void run() {
        if (password != null) {
            login(password);
        } else if (passwordEnvironmentVariable != null) {
            login(System.getenv(passwordEnvironmentVariable));
        } else if (passwordFile != null) {
            // below uses Java 8 NIO, create your own on older Java versions
            /*
            login(new String(Files.readAllBytes(passwordFile.toPath())));
            */
        } else {
            throw new ParameterException(spec.commandLine(), "Password required");
        }
    }

    private void login(String pwd) {
        System.out.printf("Password: %s%n", pwd);
    }

    public static void main(String[] args) {
        new CommandLine(new PasswordDemo()).execute(args);
    }
}

File: picocli-examples/src/main/kotlin/picocli/examples/kotlin/interactive/PasswordDemo.kt
package picocli.examples.kotlin.interactive

import picocli.CommandLine
import picocli.CommandLine.*
import picocli.CommandLine.Model.CommandSpec
import java.io.File
import java.nio.file.Files
import kotlin.system.exitProcess

class PasswordDemo : Runnable {
    @Option(names = ["--password:file"])
    var passwordFile: File? = null
    @Option(names = ["--password:env"])
    var passwordEnvironmentVariable: String? = null
    @Option(names = ["--password"], interactive = true)
    var password: String? = null
    @Spec
    var spec: CommandSpec? = null

    override fun run() {
        when {
            password != null -> {
                login(password!!)
            }
            passwordEnvironmentVariable != null -> {
                login(System.getenv(passwordEnvironmentVariable))
            }
            passwordFile != null -> {
                login(String(Files.readAllBytes(passwordFile!!.toPath())))
            }
            else -> {
                throw ParameterException(spec!!.commandLine(), "Password required")
            }
        }
    }

    private fun login(pwd: String) {
        println("Password: %s%n".format(pwd))
    }
    companion object {
        @JvmStatic fun main(args: Array<String>) {
            CommandLine(PasswordDemo()).execute(*args)
        }
    }
}
// NOTE: below is an alternative to defining a @JvmStatic main function in a companion object:
// fun main(): Unit = exitProcess(CommandLine(PasswordDemo()).execute("--password"))
